// tslint-eslint-rules/dist/rules/handleCallbackErrRule.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Lint = require('tslint');
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        var languageService = Lint.createLanguageService(sourceFile.fileName, sourceFile.getFullText());
        return this.applyWithWalker(new ErrCallbackHandlerWalker(sourceFile, this.getOptions(), languageService));
    };
    Rule.FAILURE_STRING = 'Expected error to be handled';
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var ErrCallbackHandlerWalker = (function (_super) {
    __extends(ErrCallbackHandlerWalker, _super);
    function ErrCallbackHandlerWalker(sourceFile, options, languageService) {
        _super.call(this, sourceFile, options);
        this.languageService = languageService;
        var errorArgument = options.ruleArguments[0] || 'err';
        if (errorArgument.charAt(0) === '^') {
            this.errorCheck = RegExp.prototype.test.bind(new RegExp(errorArgument));
        }
        else {
            this.errorCheck = function (name) { return name === errorArgument; };
        }
    }
    ErrCallbackHandlerWalker.prototype.visitFunctionExpression = function (node) {
        this.validateFunction(node);
        _super.prototype.visitFunctionExpression.call(this, node);
    };
    ErrCallbackHandlerWalker.prototype.visitFunctionDeclaration = function (node) {
        this.validateFunction(node);
        _super.prototype.visitFunctionDeclaration.call(this, node);
    };
    ErrCallbackHandlerWalker.prototype.visitArrowFunction = function (node) {
        this.validateFunction(node);
        _super.prototype.visitArrowFunction.call(this, node);
    };
    ErrCallbackHandlerWalker.prototype.validateFunction = function (node) {
        var parameter = node.parameters[0];
        if (parameter && this.errorCheck(parameter.name.getText())) {
            var fileName = this.getSourceFile().fileName;
            var highlights = this.languageService.getDocumentHighlights(fileName, parameter.pos, [fileName]);
            if (!highlights || highlights[0].highlightSpans.length <= 1) {
                this.addFailure(this.createFailure(parameter.name.getStart(), parameter.name.getWidth(), Rule.FAILURE_STRING));
            }
        }
    };
    return ErrCallbackHandlerWalker;
}(Lint.RuleWalker));
